api/src/websocket/controllers/base.ts

```diff
@@ -0,0 +1,246 @@
+					// ignore errors
+				}
+				return;
+			}
+			ws.emit('parsed-message', message);
+		});
+		ws.on('error', () => {
+			logger.debug(`Websocket#${client.uid} connection errored`);
+			if (client.auth_timer) {
+				clearTimeout(client.auth_timer);
+				client.auth_timer = null;
+			}
+			this.clients.delete(client);
+		});
+		ws.on('close', () => {
+			logger.debug(`Websocket#${client.uid} connection closed`);
+			if (client.auth_timer) {
+				clearTimeout(client.auth_timer);
+				client.auth_timer = null;
+			}
+			this.clients.delete(client);
+		});
+		logger.debug(
+			`Websocket#${client.uid} connected ${accountability ? JSON.stringify(accountability) : 'without authentication'}`
+		);
+		this.setTokenExpireTimer(client);
+		this.clients.add(client);
+		return client;
+	}
+	protected parseMessage(data: string): WebSocketMessage {
+		let message: WebSocketMessage;
+		try {
+			message = WebSocketMessage.parse(parseJSON(data));
+		} catch (err: any) {
+			throw new WebSocketException('server', 'INVALID_PAYLOAD', 'Unable to parse the incoming message.');
+		}
+		return message;
+	}
+	protected async handleAuthRequest(client: WebSocketClient, message: WebSocketAuthMessage) {
+		try {
+			const { accountability, expires_at, refresh_token } = await authenticateConnection(message);
+			client.accountability = accountability;
+			client.expires_at = expires_at;
+			this.setTokenExpireTimer(client);
+			emitter.emitAction('websocket.auth.success', { client });
+			client.send(authenticationSuccess(message.uid, refresh_token));
+			logger.debug(`Websocket#${client.uid} authenticated ${JSON.stringify(client.accountability)}`);
+		} catch (error) {
+			logger.debug(`Websocket#${client.uid} failed authentication`);
+			emitter.emitAction('websocket.auth.failure', { client });
+
+			client.accountability = null;
+			client.expires_at = null;
+			const _error =
+				error instanceof WebSocketException
+					? error
+					: new WebSocketException('auth', 'AUTH_FAILED', 'Authentication failed.', message.uid);
+			handleWebsocketException(client, _error, 'auth');
+
+			if (this.authentication.mode !== 'public') {
+				client.close();
+			}
+		}
+	}
+	setTokenExpireTimer(client: WebSocketClient) {
+		if (client.auth_timer !== null) {
+			clearTimeout(client.auth_timer);
+			client.auth_timer = null;
+		}
+		if (!client.expires_at) return;
+		const expiresIn = Math.max(client.expires_at * 1000 - Date.now(), 10);
+		client.auth_timer = setTimeout(() => {
+			client.accountability = null;
+			client.expires_at = null;
+			handleWebsocketException(client, new TokenExpiredException(), 'auth');
+			waitForMessageType(client, 'auth', this.authentication.timeout).catch((msg: WebSocketMessage) => {
+				const error = new WebSocketException('auth', 'AUTH_TIMEOUT', 'Authentication timed out.', msg?.uid);
+				handleWebsocketException(client, error, 'auth');
+				if (this.authentication.mode !== 'public') {
+					client.close();
+				}
+			});
+		}, expiresIn);
```

Comment:
I'm not sure what could be the safe guard yet, but it seems like when `expiresIn` is calculated to be higher than 32 bit integer limit (`2,147,483,647`), the setTimeout here will throw overflow warning and default to `1` instead:

![](https://user-images.githubusercontent.com/42867097/226635733-205b4a67-21a7-40d0-ac77-7f17a3259513.png)

which then cause said auth attempt to immediately receive TOKEN_EXPIRED error.

To reproduce this, try setting `ACCESS_TOKEN_TTL` to `25d` (because this converts to slightly higher than 32 bit int limit in milliseconds) or higher and authenticate with email/password.

Reply:
That is a very good catch, it would have never occurred to me that someone might set their `ACCESS_TOKEN_TTL` to something ridiculous like 20+days 😂 may as well just work with static access tokens at that point.

But bad security configurations aside this definitely should not overflow the integer 🤔